### ISSUE TITLE
API to set chromium preferences

### ISSUE THREAD

**Issue by harunhasdal** [cypress-io/cypress#3671]
### Current behavior:

We can't specify chromium preferences when running Cypress tests. 
This becomes an issue when tests require certain browser configuration which is different to the defaults. 

An example is to disable the popup dialog for file downloads (requires `download.prompt_for_download = true`). 

See more preferences in the link in this comment: https://github.com/cypress-io/cypress/issues/433#issuecomment-470867819

### Desired behavior:

A Cypress API to enable setting chromium preferences as key value pairs.
This could be a new API or an extension to one of the existing APIs such as the browser-launch-api or the configuration-api.

### Versions

Valid for both Chrome and Electron setups.

**Comment by jennifer-shehane** [comment]
Chromium code using prefs: chromium.googlesource.com/chromium/src/+/master/chrome/common/pref_names.cc#1427

**Comment by baptistepattyn** [comment]
Is this ready yet?

**Comment by pardonmeme** [comment]
Is it a thing? The prompt disabling is crucial ._ .

**Comment by CSchulz** [comment]
I just dived a little bit through the cypress code and in my opinion we could add a plugin event similar to `before:browser:launch` to allow passing some preferences to the profile.

It could be executed somewhere around here:
https://github.com/cypress-io/cypress/blob/a877474df1300af3dc8edb7bbe69b935a6ccc850/packages/server/lib/browsers/chrome.coffee#L281-L290

The code for reading and writing the profile can be found here:
https://github.com/cypress-io/cypress/blob/a877474df1300af3dc8edb7bbe69b935a6ccc850/packages/server/lib/browsers/chrome.coffee#L137-L146

@jennifer-shehane any thoughts on it? We can also discuss it in gitter, before I would start with a small PoC pull request.

**Comment by nmschulte** [comment]
I use this work-around currently.  I haven't tried a persistent `--user-data-dir` (a la `exec chromium $@ --user-data-dir=~/.config/cypress-chromium`, but that may work as well.

`$ npx cypress run -b ./cypress/chromium.sh`

---
`$ cat cypress/chromium.sh`
```
#!/usr/bin/env bash

USER_DATA_DIR=`cut -d "=" -f 2 <<< ${*: -2:1}`
mkdir $USER_DATA_DIR/Default > $USER_DATA_DIR/Default/Preferences
cat << END_Preferences > $USER_DATA_DIR/Default/Preferences
{
"profile":{"content_settings":{"exceptions":{"cookies":{"example.com,*":{"setting":1}}}}},
"devtools":{"preferences":{"currentDockState":"\"undocked\"","panel-selectedTab":"\"sources\"","Inspector.drawerSplitViewState":"{\"horizontal\":{\"size\":0,\"showMode\":\"Both\"}}"}},
"browser":{"app_window_placement":{"DevToolsApp":{"maiximized": true}}}
}
END_Preferences
exec chromium $@
```

**Comment by CSchulz** [comment]
@nmschulte Your way doesn't work, because cypress is creating an own user data dir during startup.

Another way would be to have access on the Chrome Debugging Protocol.

@jennifer-shehane Do you know if there is any possibility to get the current connection?

**Comment by flotwig** [comment]
@CSchulz you are able to make a connection to the Chrome DevTools Protocol in `before:browser:launch`, since you receive the `--remote-debugging-port` argument there, you know what port to use for your connection. The rest is just a matter of connecting to it and sending the commands you want.

**Comment by CSchulz** [comment]
@flotwig Thanks for your input! I will try it, but it seems Electron doesn't use anything like that, right?

**Comment by flotwig** [comment]
@CSchulz ah yes, good point, this will only work for Chrome-family browsers.

To clarify, are you asking for CDP access from your tests, or from your pluginsfile?

**Comment by CSchulz** [comment]
@flotwig I want to use it in the plugins file.

**Comment by CSchulz** [comment]
@flotwig I want to use it in the plugins file but it seems not working, because the browser is launched after the complete event handling. Using timeout seems a little bit hacky.

**Comment by flotwig** [comment]
@CSchulz I agree, it is a little bit hacky, right now that's the only way to do it unfortunately. I wrote a plugin that does a retry loop on `before:browser:launch` to connect to CDP, you can see how I did it here: https://github.com/flotwig/cypress-log-to-output/blob/master/src/log-to-output.js#L106-L142

**Comment by nmschulte** [comment]
@CSchulz it works for me; Chromium from Debian Sid.  Chromium loads the Preferences stored in the specified dir; pressing F12 loads the dev tools in a separate window.

```
Cypress:    3.6.1
Browser:    Custom Chromium 78
```
```
Version 78.0.3904.97 (Developer Build) built on Debian bullseye/sid, running on Debian bullseye/sid (64-bit)
```

@flotwig does the Chrome DevTools Protocol allow to change the browser settings?  https://stackoverflow.com/questions/57498081/chrome-devtools-protocol-interact-with-settings

**Comment by CSchulz** [comment]
@nmschulte Yes you are right, it is using the same one for interactive mode, you can see it in the process explorer.

The chrome driver overwrites the _user-data-dir_ always:

https://github.com/cypress-io/cypress/blob/a3265ccdef91ef93bc831f9a73f0628ec4564d41/packages/server/lib/browsers/chrome.coffee#L284-L308

**Comment by jennifer-shehane** [comment]
Setting Chromium preferences was released in [Cypress 4.0.0+](https://on.cypress.io/changelog#4-0-0). You can find the new documentation on how to do this in: http://on.cypress.io/browser-launch-api

@Bkucera @flotwig I'd love if one of you could provide an example of how to pass the preferences for the scenarios above. 

I'm not really sure what the original `download.prompt_for_download = true` was asking for though.

- Download prompts do not display in Chrome
- Download prompts display in Electron, but I don't see an available preference to pass to Electron to change this behavior.

**Comment by Bkucera** [comment]
see an example for setting the default download location here #6494
